README.md
# inca: an R package for integer calibration

### Authors
[Luca Sartore](mailto://luca.sartore@nass.usda.gov) and [Kelly Toppin](mailto://kelly.toppin@nass.usda.gov)

Maintainer: [Luca Sartore](mailto://drwolf85@gmail.com)

[![](http://www.r-pkg.org/badges/version/inca)](http://www.r-pkg.org/pkg/inca)
[![CRAN RStudio mirror downloads](http://cranlogs.r-pkg.org/badges/inca)](http://www.r-pkg.org/pkg/inca)

## Features of the package
Calibration forces the weighted estimates of calibration variables to match known totals. This improves the quality of the design-weighted estimates. It is used to adjust for non-response and/or under-coverage. The commonly used methods of calibration produce non-integer weights. In cases where weighted estimates must be integers, one must "integerize" the calibrated weights. However, this procedure often produces final weights that are very different for the "sample" weights. To counter this problem, the *inca* package provides specific functions for rounding real weights to integers, and performing an integer programming algorithm for calibration problems with integer weights.

For a complete list of exported functions, use `library(help = "inca")` once the *inca* package is installed (see the `inst/INSTALL.md` file for a detailed description of the setup process).

### Example
```R
library(inca)
set.seed(0)
w <- rpois(150, 4)
data <- matrix(rbinom(150000, 1, .3) * rpois(150000, 4), 1000, 150)
y <- data %*% w
w <- runif(150, 0, 7.5)
print(sum(abs(y - data %*% w)))
cw <- intcalibrate(w, ~. + 0, y, lower = 1, upper = 7, sparse = TRUE, data = data)
print(sum(abs(y - data %*% cw)))
barplot(table(cw), main = "Calibrated integer weights")
```

## References
Theberge, A. (1999). Extensions of calibration estimators in survey sampling. Journal of the American Statistical Association, 94(446), 635-644.

Little, R. J., & Vartivarian, S. (2003). On weighting the rates in non-response weights.

Kish, L. (1992). Weighting for unequal Pi. Journal of Official Statistics, 8(2), 183.

Rao, J. N. K., & Singh, A. C. (1997). A ridge-shrinkage method for range-restricted weight calibration in survey sampling. In Proceedings of the section on survey research methods (pp. 57-65). American Statistical Association Washington, DC.

Horvitz, D. G., & Thompson, D. J. (1952). A generalization of sampling without replacement from a finite universe. Journal of the American Statistical Association, 47(260), 663-685.

Kalton, G., & Flores-Cervantes, I. (2003). Weighting methods. Journal of Official Statistics, 19(2), 81-98.



ChangeLog
2018-02-09 Luca Sartore <drwolf85@gmail.com>
	* DESCRIPTION (Version): New version is 0.0.3
	* DESCRIPTION (Date): New date is 2018-02-09
	* inca.R: All references to the new version and date are now
           changed to match those in the DESCRIPTION file
	* inst/INSTALL.md: All references to the new version are now
           changed to match those in the DESCRIPTION file
	* src/inca.cpp: The new routine `R_init_inca` is now added
	       to register routines and improving the use of dynamic
	       symbols
	* inca.R: It was updated the `@useDynLib` to allowed for the
	       registration of routines
	* README.md: A duplicated reference was removed

2016-09-14 Luca Sartore <drwolf85@gmail.com>
	* DESCRIPTION (Version): New version is 0.0.2
	* DESCRIPTION (Date): New date is 2016-09-14
	* inca.R: All references to the new version and date are now
           changed to match those in the DESCRIPTION file
	* inst/INSTALL.md: All references to the new version are now
           changed to match those in the DESCRIPTION file
	* src/Losses/aL1.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/Losses/L1.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/Losses/LB1.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/Losses/rB1.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/Losses/rL1.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/bestRound.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* src/intCalib.h: The armadillo `sort_index()` function is
	    now replaced by `stable_sort_index()`
	* THANKS: Included a further contributor who discovered a bug



R/inca.R
#' @importFrom stats model.matrix.lm
#' @importFrom Matrix Matrix
#' @importFrom Rcpp evalCpp
#' @useDynLib inca, .registration = TRUE
#'
#' @name inca-package
#' @aliases inca
#' 
#' @docType package
#' 
#' @title
#' Integer Calibration
#' 
#' @description
#' Specific functions are provided for rounding real weights to integers and performing an integer programming algorithm for calibration problems
#' 
#' @details
#' \tabular{ll}{
#' Package: \tab inca\cr
#' Type: \tab Package\cr
#' Version: \tab 2018-02-09\cr
#' Date: \tab 0.0.3\cr
#' License: \tab GPL (>= 2)\cr
#' }
#' 
#' Calibration forces the weighted estimates of calibration variables to match known totals. 
#' This improves the quality of the design-weighted estimates. It is used to adjust for 
#' non-response and/or under-coverage. The commonly used methods of calibration produce 
#' non-integer weights. In cases where weighted estimates must be integers, one must 
#' "integerize" the calibrated weights. However, this procedure often produces final weights 
#' that are very different for the "sample" weights. To counter this problem, the *inca* 
#' package provides specific functions for rounding real weights to integers, and performing 
#' an integer programming algorithm for calibration problems with integer weights.
#'
#' For a complete list of exported functions, use \code{library(help = "inca")}.
#' 
#' @author
#' Luca Sartore \email{luca.sartore@@nass.usda.gov} and Kelly Toppin \email{kelly.toppin@@nass.usda.gov}
#'
#' Maintainer: Luca Sartore \email{drwolf85@@gmail.com}
#' 
#' @references
#' Theberge, A. (1999). Extensions of calibration estimators in survey sampling. Journal of the American Statistical Association, 94(446), 635-644.
#'
#'
#' Little, R. J., & Vartivarian, S. (2003). On weighting the rates in non-response weights.
#' Kish, L. (1992). Weighting for unequal Pi. Journal of Official Statistics, 8(2), 183.
#'
#' Rao, J. N. K., & Singh, A. C. (1997). A ridge-shrinkage method for range-restricted weight calibration in survey sampling. In Proceedings of the section on survey research methods (pp. 57-65). American Statistical Association Washington, DC.
#'
#' Horvitz, D. G., & Thompson, D. J. (1952). A generalization of sampling without replacement from a finite universe. Journal of the American Statistical Association, 47(260), 663-685.
#'
#' Horvitz, D. G., & Thompson, D. J. (1952). A generalization of sampling without replacement from a finite universe. Journal of the American Statistical Association, 47(260), 663-685.
#'
#' Kalton, G., & Flores-Cervantes, I. (2003). Weighting methods. Journal of Official Statistics, 19(2), 81-98.
#'
#' @keywords
#' integer calibration rounding
#' 
#' @examples
#' library(inca)
#' 
NULL
